keps/sig-scheduling/5471-enable-sla-based-scheduling/README.md

```diff
@@ -0,0 +1,1430 @@
+- Consistent semantic version handling across validation policies
+
+This creates a unified approach to threshold-based and version-based placement/validation across Kubernetes, preserving the well‑understood safety model of taints/tolerations while enabling sophisticated policy enforcement through CEL expressions.
+
+## Motivation
+
+Many clusters blend (**on‑demand/higher‑SLA**) and (**spot-preemptible/lower‑SLA**) nodes. Platform teams want a safe default keeping most workloads away from risky capacity, while allowing specific workloads to opt‑in with explicit thresholds like `SLA ≥ 95%`.
+
+### Why not NodeAffinity alone?
+
+For the “node SLA / failure‑probability” use‑case, NodeAffinity can express minimum or exact SLA thresholds via label comparisons, but it’s not sufficient for the operational goals here:
+
+- **Policy orientation:** NodeAffinity is per‑pod; to keep most pods away from low‑SLA nodes you'd have to edit every workload.
+- **Taints invert control**: nodes declare risk; only pods with a matching toleration may land.
+- **Eviction semantics:** Affinity has no eviction. Taints support `NoExecute` with `tolerationSeconds`, letting operators drain/evict pods when a node's SLA class drops or a spot reclaim hits.
+- **Operational ergonomics:** Centralized, node‑side policy is consistent with other safety taints (e.g., disk-pressure, memory-pressure). Teams opt‑in, reducing config drift.
+
+From a scheduling perspective, adding numeric operators to tolerations only adjusts match logic. It does not change queueing, scoring, or preemption algorithms.
+
+### Goals
+
+- Add comparison operators to tolerations so pods can match taints like `node.kubernetes.io/sla=<int>` using thresholds.
+- Enable automatic type detection for integer and semantic version comparisons in tolerations.
+- Enhance CEL semver library with additional comparison functions for consistent version handling.
+- Keep behavior consistent with existing effects (`NoSchedule`, `PreferNoSchedule`, `NoExecute`).
+- Provide unified semantic version handling across scheduling and admission control.
+- Backward compatible and opt‑in via feature gates.
+
+### Non-Goals
+
+- Standardizing an SLA key or unit (clusters may choose any integer scale, e.g., 950 for 95.0%).
+- Implementing workload‑level "70/30" mix semantics.
+- Changing NodeAffinity behavior.
+- Replacing existing CEL semver functions (maintaining backward compatibility).
+- Adding new taint/toleration keys (focusing on operator extensions only).
+
+### Benefits for implementing this feature for DRA and AI Workloads
+
+In addition to general scheduling improvements, SLA‑aware opt‑in via tolerations has specific advantages for `Dynamic Resource Allocation (DRA)` and `AI/ML`:
+
+- DRA steers GPUs/accelerators resource claims by node reliability: critical workloads get high‑SLA capacity while batch workloads use cheaper pools. Taints block risky pools and evict when capacity degrades.
```

Comment:
nit: maybe we add "interruptible batch workloads" or checkpoint-able or other similar language to recognize that there is such a thing as a "critical batch workload" that would not be able to benefit from running on interruptible hardware.

Reply:
Addressed